cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (\*)
+2. Never change the language syntax using `#deﬁne`.
+3. Do not use spaces between a function, constructor, or method name and its opening parenthesis, e.g. `foo()` rather than `foo ()`. A similar style is encouraged but not required when brace initialization is used, e.g. `foo{13}`.
+4. Do not use spaces in front of [] or on either side of -> . For example, `vector[i]` instead of `vector [i]`.
+5. Separate expressions in a `for` statement by spaces.
+6. Use the same indentation for comments as for the block the comments refer to.
+
+## 4 -- Technical Coding Rules
+1. Protect each header ﬁle from multiple inclusion with:  
+`#ifndef PackageName_SubPackageName_FileName_h`  
+`#define PackageName_SubPackageName_FileName_h`  
+(body of header file)  
+`#endif`  
+If necessary to create a unique name, one can add the directory name:  
+`PackageName_SubPackageName_Directory_FileName_h`.
+2. Each header ﬁle contains one class declaration only. (\*)
+3. Header ﬁles must not contain any implementation except for class templates and code to be inlined.
+4. Do not inline virtual functions.
+5. Do not inline functions which contain control structures which require block scoping.
+6. In your own packages, use forward declarations if they are sufﬁcient.
+7. Do not forward-declare an entity from another package.
+8. Do not use absolute directory names or relative ﬁle paths in `#include` directives.
+9. Use `nullptr`, not "0" or "NULL".
+10. Use types like `int`, `uint32_t`, `size_t`, and `ptrdiff_t` consistently and without mixing them.
+11. Use the `bool` type for booleans.
+12. Copy and move assignment operators should return a reference to `*this`.
+13. For a class, definition of any of the following requires definition of all five: destructor, copy constructor, copy assignment operator, move constructor, and move assignment operator. (\*)
+14. Do not use function-like macros.
+15. Use C++ casts, not C-style casting. (\*)
+16. Do not use the ellipsis notation for function arguments, except for variable argument templates. (\*)
+17. Do not use union types. (\*)
+18. If a class has at least one virtual method, it must have a public virtual destructor or (exceptionally) a protected destructor.
+19. When a derived class function overrides a virtual function, always mark it with `override` or `final`.
+20. Pass by value arguments which are not to be modiﬁed and are built-in types or small objects; otherwise pass arguments of class types by reference or, if necessary, by pointer.
+21. Properly use rvalue references for temporary objects that will be moved.
+22. The argument to a copy constructor and to an assignment operator must be a `const` reference, while the argument for a move constructor or move assignment operator must be an rvalue reference. (\*)
+23. Do not let `const` member functions change the state of the object. (\*)
```

Comment:
Regarding the exception, should we note that even if a member function would change the state, it must be thread-safe?

Reply:
Added mention of exception.